### DISCUSSION TITLE
Custom MCCS file not generating expected results

### DISCUSSION THREAD

**Discussion by ventsyv** [rockowitz/ddcutil#435] [Q&A]
I created a file name `DEL-DELL_U3821DW-41390.mccs` in `~/.local/share/ddcutil`
The content is as follows:

```
MFG_ID       DEL
MODEL        DELL U3812DW
PRODUCT_CODE 41390
MCCS_VERSION 2.1
FEATURE_CODE 60 Input Source
	ATTRS NC RW
	VALUE 1b USB-C
```

When I run `ddcutil capabilities --verbose` I get the following output

```
Feature: 60 (Input Source)
      Values (unparsed):  1B 0F 11 12
      Values (  parsed):
         1b: Unrecognized value
         0f: DisplayPort-1
         11: HDMI-1
         12: HDMI-2
```

I expected that the "1b: Unrecognized value" to be replaced which `1b: USB_C` - the value I defined in the MCCS file.
I tried using hex 0x1b also, and that did not change the output.

Am I mis-understanding that the custom mccs will override any specified settings  with values from the file?

The rest of the input source value match by the way.

**Comment by rockowitz** [comment]
There are several things going on here.

I expect that at the start of the capabilities output there were messages reporting that the UDF file is invalid. As a result, it was ignored.  The source in branch 2.1.5-dev has been changed to terminate execution in this situation, which should make the failure clearer. 

Why was your UDF file invalid?
- It may just be an artefact off the text posted here, but there are leading tabs on some lines instead of spaces.  In branch 2.1.5-dev, the UDF parser has been modified to recognize any whitespace characters, not just spaces.
- More significantly, "VALUE 1b USB-C" needs to be written as "VALUE x1b USB-C'.  The rule for parsing values here, as on **setvcp**, is to regard a value as decimal unless explicitly marked as hex. This doesn't matter for values <= x09, but disambiguation is needed for values >= x10. 

Regarding the latter point, the **capabilities** command doesn't make clear that the values reported are hex values.
That is, "11" means x11, i.e. decimal 17, not decimal 11.    This needs to be better documented.  I will also consider whether to change the **capabilities** output to make this clearer.  Unfortunately, this could break screen scraping apps.  There's no guarantee, except in specific documented cases, that the format and content of messages intended for human consumption are fixed, but as a practical matter I do try to avoid changes that can impact those programs.

**Comment by ventsyv** [comment]
Thanks for the feedback, however:
1. The tabs are definitely an artifact of the copy-paste process. I don't have tabs in the file.
2. I had the model number transposed. Fixing that and using a hex value for the input value fixed the parsing errors.

However, the output is still not as I expect it to be. The 1b input value is still showing as "unrecognized".

I'm attaching my mccs file and the output of the capabilities command.
[capabilities.txt](https://github.com/user-attachments/files/16333956/capabilities.txt)
[DEL-DELL_U3821DW-41390.mccs.txt](https://github.com/user-attachments/files/16334010/DEL-DELL_U3821DW-41390.mccs.txt)

I'm running `ddcutil 1.4.1` which is the default package that Ubuntu installs.

**Reply by ventsyv** [reply]
Also, with that MCCS file is in my `~/.local/shared/ddcutil/` directory, when I run `ddcutil setvcp 60 0x1b` I get the following back:
`Feature 0x60 (Input Source) is not writable`

**Comment by rockowitz** [comment]
Your mccs file is missing line "ATTRS NC RW".  Also, the only VALUE line is for feature 0x1b.  No other values are listed. The feature definition in the mccs file completely overrides the default feature definition (from the mccs spec).

In case you're not already doing so, use option ***--verbose***.   There will be much more information 

Release 1.4.1 is a year and a half old.   Among other things the latest version in branch 2.1.5-dev has much clearer reports of how the feature definition file is processed.  Beyond a certain point I can't help you much if you're on that very old version. If you're still having problems, please build from branch 2.1.5-dev.

**Reply by ventsyv** [reply]
Fair enough. Thanks!
